### PR TITLE
Catch optimizer panics

### PR DESCRIPTION
**Pull request by timvisee**
Catch any panics happening during optimization and handle this situation.

Now, when an optimization task panics it is reported and the collection status is marked as `red`. Before, a collection hung as `yellow` indefinitely. Now:

image

### Explanation

In short, this now catches panics as they unwind in the optimization threads. When an optimization task is done, it is now explicitly joined upon to nicely clean them up. On joining, any panic that occurred is collected. These panics are then propagated through a panic handling callback. This new callback is attached to our stoppable task structure.

So a stoppable task may now define a panic callback to handle the panic status. These stoppable tasks are used for optimization. Our optimization tasks now define a panic callback to report it, registering the error and setting the collection state to red.

We are not notified about a panicked task. Our update handler therefore periodically checks for finished (and potentially panicked) optimization tasks. It cleans them up and propagates any panic. This happens at a fixed interval of 5 seconds. So in the worst case, it may take a few seconds for the red collection status to appear.

I've rewritten the way we use our optimizer worker and task handling a bit to make all of this happen.

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: lib/collection/src/update_handler.rs [hunk 1]
```diff
@@ -213,42 +220,59 @@ impl UpdateHandler {
...
+                                        panic!("Optimization error: {error}");
+                                    }
+                                },
+                            }
                         }
-                    }));
-                }
+                    },
+                    // Panic handler
+                    Some(Box::new(move |panic_payload| {
```
Comment (timvisee) [inline lib/collection/src/update_handler.rs:263]: Pointer: this new callback now gets any panic payload and reports the error on the segment holder.

File: lib/collection/src/update_handler.rs [hunk 2]
```diff
@@ -285,18 +335,27 @@ impl UpdateHandler {
...
     ) {
-        while let Some(signal) = receiver.recv().await {
-            match signal {
-                OptimizerSignal::Nop | OptimizerSignal::Operation(_) => {
+        loop {
+            let receiver = timeout(OPTIMIZER_CLEANUP_INTERVAL, receiver.recv());
+            let result = receiver.await;
+
+            // Always clean up on any signal
+            Self::cleanup_optimization_handles(optimization_handles.clone()).await;
```
Comment (timvisee) [inline lib/collection/src/update_handler.rs:343]: Pointer: the worker now waits for any signal, but also continues the loop every 5 seconds.

This way we always clean up finished optimization tasks at least every 5 seconds, as that is done right after. This is needed to catch and propagate panics in time.

If we actually received a signal, it is handled below.

File: lib/collection/src/update_handler.rs [hunk 3]
```diff
@@ -213,42 +220,59 @@ impl UpdateHandler {
...
+                        let panic_msg = panic_payload_into_string(panic_payload);
+                        log::warn!(
+                            "Optimization task panicked, collection may be in unstable state: {panic_msg}"
+                        );
+                        segments
+                            .write()
+                            .report_optimizer_error(CollectionError::service_error(format!(
+                                "Optimization task panicked: {panic_msg}"
+                            )));
+                    })),
```
Comment (timvisee) [inline lib/collection/src/update_handler.rs:273]: Pointer: this new callback now gets any panic payload and reports the error on the segment holder.

File: lib/collection/src/update_handler.rs (hunk 2 above)
Comment (timvisee) [inline lib/collection/src/update_handler.rs:343]: Pointer: the worker now waits for any signal, but also continues the loop every 5 seconds.

This way we always clean up finished optimization tasks at least every 5 seconds, as that is done right after. This is needed to catch and propagate panics in time.

If we actually received a signal, it is handled below.

File: lib/collection/src/update_handler.rs [hunk 4]
```diff
@@ -285,18 +335,27 @@ impl UpdateHandler {
         optimization_handles: Arc<TokioMutex<Vec<StoppableTaskHandle<bool>>>>,
         max_handles: usize,
     ) {
-        while let Some(signal) = receiver.recv().await {
-            match signal {
-                OptimizerSignal::Nop | OptimizerSignal::Operation(_) => {
+        loop {
```
Comment (timvisee) [inline lib/collection/src/update_handler.rs:338]: Maybe review this loop with extra care as I've restructured the logic a bit.

File: lib/collection/src/common/stoppable_task.rs [hunk 5]
```diff
@@ -1,12 +1,16 @@
...
 
 use tokio::task::JoinHandle;
 
+type PanicPayload = Box<dyn Any + Send + 'static>;
+
 pub struct StoppableTaskHandle<T> {
     pub join_handle: JoinHandle<Option<T>>,
     started: Arc<AtomicBool>,
     stopped: Weak<AtomicBool>,
+    panic_handler: Option<Box<dyn Fn(PanicPayload) + Sync + Send>>,
```
Comment (agourlay) [inline lib/collection/src/common/stoppable_task.rs:13]: Why not making the `panic_handler` required?
Would that make the PR too large due to all the call sites?

File: lib/collection/src/common/stoppable_task.rs [hunk 6]
```diff
@@ -28,9 +32,48 @@ impl<T> StoppableTaskHandle<T> {
...
         self.is_started().then_some(self.join_handle)
     }
+
+    /// Join this stoppable task and handle any panics
+    ///
+    /// Any panics are propagated through the configured panic handler. If no handler is
+    /// configured, nothing happens.
+    ///
+    /// To call this, the task must already be finished. Otherwise it panics in development, or
+    /// blocks in release.
```
Comment (agourlay) [inline lib/collection/src/common/stoppable_task.rs:42]: Could have have a large timeout instead of blocking forever?

Comment (timvisee) [reply lib/collection/src/common/stoppable_task.rs:42]: Yeah. But this method should not be called before the task is finished anyway. That's why there's a (debug) assert here to ensure that.

File: lib/collection/src/common/stoppable_task.rs (hunk 5 above)
Comment (timvisee) [reply lib/collection/src/common/stoppable_task.rs:13]: Could do. I'm not afraid of it becoming too large. But most of the calls will just have an empty handler.

Also, regular threads don't require (or support) a panic handler either. Most of the time, these will just panic silently.

File: lib/collection/src/update_handler.rs [hunk 7]
```diff
@@ -285,18 +335,27 @@ impl UpdateHandler {
...
+            let result = receiver.await;
+
+            // Always clean up on any signal
+            Self::cleanup_optimization_handles(optimization_handles.clone()).await;
+
+            match result {
+                // Channel closed or stop signal
+                Ok(None | Some(OptimizerSignal::Stop)) => break,
+                // Clean up interval
+                Err(_) => continue,
```
Comment (agourlay) [inline lib/collection/src/update_handler.rs:349]: What about using the full type like `Err(Elapsed)`?
It would make it clearer to the reader which error we are handling.
It might also prevent some catch all errors bug in a future refactoring.

Comment (timvisee) [reply lib/collection/src/update_handler.rs:349]: Good idea. Updating.

File: lib/collection/src/common/stoppable_task.rs (hunk 5 above)
Comment (agourlay) [reply lib/collection/src/common/stoppable_task.rs:13]: Let's keep it like that :+1:

File: lib/collection/src/update_handler.rs (hunk 2 above)
Comment (generall) [reply lib/collection/src/update_handler.rs:344]: my concern here is that current behaviour is so, that if optimisation caused oomkill, it won't be automatically started after the restart, potentially preventing the crash loop. This might be sub-optimal in some cases, as the downside is that optimisation might appear not starting without a reason

Comment (timvisee) [reply lib/collection/src/update_handler.rs:344]: It is not entirely clear to me what you mean. Do you mean that this may catch an OOM panic without restarting?

Doesn't an OOM event termiate the process immediately without unwinding?

Comment (generall) [reply lib/collection/src/update_handler.rs:344]: What I mean, that starting optimization right after restart may cause crashloop

Comment (timvisee) [reply lib/collection/src/update_handler.rs:371]: Thanks for elaborating. I don't think it immediately starts optimization.

The signal receiver with timeout does not start optimization when it times out. Instead it only cleans up finished optimization handles (which would be none) and then waits for a signal [again](https://github.com/qdrant/qdrant/blob/d8d2306a9b176e78c8e48d720d2874938db1243d/lib/collection/src/update_handler.rs#L376-L377).

In other words, an optimization is [only](https://github.com/qdrant/qdrant/blob/d8d2306a9b176e78c8e48d720d2874938db1243d/lib/collection/src/update_handler.rs#L378-L401) started if a `OptimizerSignal::Nop` or `OptimizerSignal::Operation(_)` is received through the channel.

Or am I missing something important here?

Comment (generall) [reply lib/collection/src/update_handler.rs:371]: I can see it now, thanks! I think my concern is resolved then

File: lib/collection/src/common/stoppable_task.rs [hunk 8]
```diff
@@ -28,9 +32,48 @@ impl<T> StoppableTaskHandle<T> {
...
+    }
 }
 
-pub fn spawn_stoppable<F, T>(f: F) -> StoppableTaskHandle<T>
+/// Spawn stoppable task `f`
+///
+/// An optional `panic_handler` may be given, eventually called if the task panicked.
+pub fn spawn_stoppable<F, T>(
+    f: F,
+    panic_handler: Option<Box<dyn Fn(PanicPayload) + Sync + Send>>,
```
Comment (generall) [inline lib/collection/src/common/stoppable_task.rs:75]: nitpick: I think we need to adopt builder pattern more in future. There are some places which getting more and more arguments and each new argument involve a change in dozen of tests which doesn't use the param.
But in this concrete example it looks ok for now

**Review by agourlay — APPROVED** [review]

File: lib/collection/src/common/stoppable_task.rs (hunk 8 above)
Comment (timvisee) [reply lib/collection/src/common/stoppable_task.rs:75]: Agree. Good practice to pick up in the future.
